### PR TITLE
Make GHA Map-JPA base testsuite running with Quarkus

### LINKED ISSUE #13725: Make GHA Map-JPA base testsuite running with Quarkus
### Description

It'd be good to execute our base testsuite with the Map-JPA store.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #13725 

The #13642 and #13721 must be merged prior to this.

### CODE REVIEW

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks @mabartos, I didn't know there was a possibility to run PostgreSQL from maven.

[This](https://github.com/keycloak/keycloak/blob/1445646e77a787cdee3d9fee73b8e51383bbf0d2/.github/workflows/ci.yml#L206) should be removed if we run map-jpa tests in the Base testsuite job. 

Also, map-jpa tests should not run for all PRs, see [this](https://github.com/keycloak/keycloak/blob/1445646e77a787cdee3d9fee73b8e51383bbf0d2/.github/workflows/ci.yml#L241) condition for current map-jpa job. There is already something like [this for HotRod](https://github.com/keycloak/keycloak/blob/1445646e77a787cdee3d9fee73b8e51383bbf0d2/.github/workflows/ci.yml#L148) in this job

**Comment by mabartos** [comment]
@mhajas Thanks for your comment. I've modified the code. 

It's necessary to merge this after #13643. 

I've executed GH actions with it and everything works as expected. See #13802 GH actions.

**Comment by mabartos** [comment]
Needs to be fixed. Not a proper solution.

**Comment by mabartos** [comment]
@mhajas GH actions work as expected. The Map-JPA and Map-HotRod tests are skipped, when there's no change. 

I've executed test GH actions for testing executing only Map-JPA. Works as expected as well. See #13802

**Review by mhajas — APPROVED** [review]
Thanks @mabartos, from my point of view this change makes sense. However, it would probably make sense to wait for someone from JPA part of the storage team to have a look. I asked in our team chat.

**Review by mhajas — CHANGES_REQUESTED** [review]
I missed one important thing during my review. I think it was there before but as I am looking again, setting for `PARAMS["quarkus-map-jpa"]` is missing. 

Should there be something like:
```
PARAMS["quarkus-map-jpa"]="-Pauth-server-quarkus -Pmap-storage,map-storage-jpa,db-postgres -Dpageload.timeout=90000"
``` 
?

**Comment by mabartos** [comment]
@mhajas Good catch! Fixed

**Review by mhajas — DISMISSED** [review]
Thanks again @mabartos

**Comment by mabartos** [comment]
@mhajas Thanks for the review!

**Review by sguilhen — COMMENTED** [review]
To be honest it took me a while to understand all these conditions, only after discussing with @mhajas and @ahus1 I was able to fully understand how they work together to run the appropriate tests. Isn't there a way to make these conditions clearer?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -144,16 +144,30 @@ jobs:
...
+        if: ${{ !endsWith(matrix.server, '-map-jpa') || env.GIT_MAP_JPA_RELEVANT_DIFF != 0 }}
+        run: echo "CHECK_MAP_JPA_EXECUTION=true" >> $GITHUB_ENV
+
+      - name: Check whether those phases should be executed
+        run: echo "SHOULD_BE_EXECUTED=$( test "$CHECK_HOTROD_EXECUTION" = true -a "$CHECK_MAP_JPA_EXECUTION" = true && echo true || echo false )" >> $GITHUB_ENV
+
       - name: Cache Maven packages
-        if: ${{ github.event_name != 'pull_request' || matrix.server != 'undertow-map-hot-rod' || env.GIT_HOTROD_RELEVANT_DIFF != 0 }}
+        if: ${{ github.event_name != 'pull_request' || env.SHOULD_BE_EXECUTED == 'true' }}
         uses: actions/cache@v3
```
Comment (ahus1) [inline .github/workflows/ci.yml:163]: I had a hard time reading this code with the negations and, to understand the final condition `SHOULD_BE_EXECUTED`.

See below for an attempt to remove that last condition, and use `env.SHOULD_BE_EXECUTED != 'false'` for the tests (would need to be applied for all other steps as well). 

```
      - name: Check whether HEAD^ contains HotRod storage relevant changes
        run: echo "GIT_HOTROD_RELEVANT_DIFF=$( git diff --name-only HEAD^ | egrep -ic -e '^model/map-hot-rod|^model/map/|^model/build-processor' )" >> $GITHUB_ENV

      - name: Check whether HotRod storage matrix should be executed
        if: ${{ endsWith(matrix.server, '-map-hot-rod') && env.GIT_HOTROD_RELEVANT_DIFF == 0 }}
        run: echo "SHOULD_BE_EXECUTED=false" >> $GITHUB_ENV

      - name: Check whether HEAD^ contains JPA map storage relevant changes
        run: echo "GIT_MAP_JPA_RELEVANT_DIFF=$( git diff --name-only HEAD^ | egrep -ic -e '^model/map-jpa/|^model/map/|^model/build-processor' )" >> $GITHUB_ENV

      - name: Check whether Map-JPA storage matrix should be executed
        if: ${{ endsWith(matrix.server, '-map-jpa') && env.GIT_MAP_JPA_RELEVANT_DIFF == 0 }}
        run: echo "SHOULD_BE_EXECUTED=false" >> $GITHUB_ENV

      - name: Cache Maven packages
        if: ${{ github.event_name != 'pull_request' || env.SHOULD_BE_EXECUTED != 'false' }}
        uses: actions/cache@v3     
```

**Review by ahus1 — COMMENTED** [review]
Please see my comment below. I's only comment, neither an approve/reject. Please let me know what you think.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mabartos) [reply .github/workflows/ci.yml:163]: @sguilhen @ahus1 Thanks for your comments! Yep, I'd say this approach is more readable. 

I wasn't sure about not defined env var, but it never will be evaluated to the false (even to the string format). It should work. Thanks!

**Review by vramik — COMMENTED** [review]
Thank you @mabartos! I'd like to be sure that newly added checks works as expected. I was looking at https://github.com/keycloak/keycloak/pull/13802/checks and also at https://github.com/mabartos/keycloak/runs/7919274057?check_suite_focus=true and at both cases group2 failed with some test failures and group3 timeouted. 

Could you please check it to see whether it is related?

**Comment by mabartos** [comment]
ATM, it's not possible to execute it with the specified DB profile as the base tests are divided into 3 groups. The first group is executed properly. In the second one, some failures are likely caused by using the same not-empty DB as for group one. For the third group, the connection to the DB is not even established. 

Therefore, I assume the parallel test execution in the GH actions is done on the same node and not separately. So, we'd probably need to completely isolate the execution of the particular base test groups. It means that we could start 3 DB instances on different ports, and for each group, specify on which port the DB is available. 

Or maybe TestContainers provides some functionality that could solve this problem, but I'm not aware of it.

**Comment by vramik** [comment]
@mabartos just FYI, there is (draft) PR https://github.com/keycloak/keycloak/pull/14926 for running model tests in GHA. There is used the Testcontainers (as @mhajas pointed out above) for postgres. Maybe it could be useful.

**Comment by vramik** [comment]
Closing in favor of https://github.com/keycloak/keycloak/pull/15418

**Comment by mabartos** [comment]
@vramik Great! Thank you!

**Pull request closed without merging**